### PR TITLE
storage: add OriginID to MVCCValueHeader

### PR DESCRIPTION
**Pull request by msbutler**
This patch adds a new OriginID field to the MVCCValueHeader which identifies the original cluster that wrote the key in a Logical Data Replication (LDR) setup. A value of 0 identifies a local write, a value of 1 identifies a remote write, and values above 2+ may be used eventually to identify a specific cluster.

In a future patch, the OriginID field will be used in a similar fashion to the OmitInRangefeeds MVCCValueHeader field -- downstream rangefeeds may be configured to filter logical operations depending on the origin ID. This approach will address the LDR data looping problem. For more information, read in the linked epic:

Epic: https://cockroachlabs.atlassian.net/browse/CRDB-39140

Release note: none

### CODE REVIEW

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126118#-:-O0BEJG_9dX5FIuu57dm:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt, @jbowens, @msbutler, and @nvanbenschoten)

___
*[`pkg/storage/enginepb/mvcc3.proto` line 212 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126118#-O0BE1u24SxaKT0Xpyup:-O0BE1u24SxaKT0Xpyuq:b-h82ji5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa89d60de85602744b70fdb52b69f841ca4e4a4c/pkg/storage/enginepb/mvcc3.proto#L212)):*
> ```proto
>   
>   // OriginID identifes the original cluster that wrote this key in Logical Data
>   // Replication. 0 identifies a local write, 1 identifies a remote write, and
> ```

[nit] maybe mention that an omitted field is the same as a local write?

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126118)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt, @jbowens, and @nvanbenschoten)

___
*[`pkg/storage/enginepb/mvcc3.proto` line 212 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126118#-O0BE1u24SxaKT0Xpyup:-O0BGmhv2hFzIu6-oWO1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa89d60de85602744b70fdb52b69f841ca4e4a4c/pkg/storage/enginepb/mvcc3.proto#L212)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] maybe mention that an omitted field is the same as a local write?
</blockquote></details>

Done.

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> Reviewed 10 of 11 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126118#-:-O0BHcqa41J93yfDM4af:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt, @jbowens, and @nvanbenschoten)

File: pkg/storage/enginepb/mvcc3.proto [hunk 1]
```diff
@@ -207,6 +207,12 @@ message MVCCValueHeader {
   // ImportEpoch identifies the number of times a user has called IMPORT
   // INTO on the table this key belongs to when the table was not empty.
   uint32 import_epoch = 4;
+  
+  // OriginID identifes the original cluster that wrote this key during Logical
+  // Data Replication (LDR). 0 identifies a local write and the default value
+  // when LDR is not being run, 1 identifies a remote write, and 2+ are reserved
+  // to identify remote clusters.
+  uint32 origin_id = 5  [(gogoproto.customname) = "OriginID"];
 }
```
Comment (dt) [inline pkg/storage/enginepb/mvcc3.proto:216]: we should `// Next ID` these in the next PR (for timestamps)

**Review by dt — APPROVED** [review]

**Comment by msbutler** [comment]
@RaduBerinde as a heads up, I added `OriginID` to `MVCCWriteOptions` and refactored the OmitRangefeeds test to test originID as well.

**Review by RaduBerinde — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126118#-:-O0EpE1s4bSpZHt7ns6M:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens, @msbutler, and @nvanbenschoten)

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 10 of 11 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126118#-:-O0EqZ7R6uKCY99Hg-mX:bnfp4nl)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @jbowens and @msbutler)

___
*[`pkg/storage/mvcc_test.go` line 420 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126118#-O0EplcB08W8NpQsyijp:-O0EplcB08W8NpQsyijq:b-e9os5j) ([raw file](https://github.com/cockroachdb/cockroach/blob/526907c0451f77f64aaaa3be53e180e3cd92c599/pkg/storage/mvcc_test.go#L420)):*
> ```Go
> 				require.NoError(t, err)
> 
> 				valueRes, vh, err := MVCCGetWithValueHeader(ctx, engine, testKey1, hlc.Timestamp{WallTime: 4}, MVCCGetOptions{})
> ```

It's important that the default case for `MVCCValueHeader` remains empty, so that we can omit it entirely from the encoding (see "simple value encoding"). What do you think about asserting that in this test? Something like:
```go
// up top
expSimpleEnc := !omitInRangefeeds && originID == 0

...

// in the assertions block after each call to MVCCGetWithValueHeader
require.Equal(t, expSimpleEnc, vh.IsEmpty())
```

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126118)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @jbowens, @nvanbenschoten, and @RaduBerinde)

___
*[`pkg/storage/mvcc_test.go` line 420 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126118#-O0EplcB08W8NpQsyijp:-O0EuN182wlz103-chKq:brqos9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/526907c0451f77f64aaaa3be53e180e3cd92c599/pkg/storage/mvcc_test.go#L420)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's important that the default case for `MVCCValueHeader` remains empty, so that we can omit it entirely from the encoding (see "simple value encoding"). What do you think about asserting that in this test? Something like:
```go
// up top
expSimpleEnc := !omitInRangefeeds && originID == 0

...

// in the assertions block after each call to MVCCGetWithValueHeader
require.Equal(t, expSimpleEnc, vh.IsEmpty())
```
</blockquote></details>

good callout. Looks like there's already a test (TestMVCCValueHeader_IsEmpty) dedicated to asserting this, so I added to it.

File: pkg/storage/enginepb/mvcc3.proto (hunk 1 above)
Comment (msbutler) [reply pkg/storage/enginepb/mvcc3.proto:218]: Done.

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126118#-:-O0F0srx6VifqJYOfHX4:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @dt and @jbowens)

**Comment by msbutler** [comment]
TFTRs!

bors r+

**Pull request merged**